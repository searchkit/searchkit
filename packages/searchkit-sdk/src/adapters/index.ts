import { SearchkitConfig } from '..'

export interface SearchkitAdapterConstructor {
  new (config: SearchkitConfig): SearchkitAdapter
}

export interface SearchkitAdapter {
  performRequest(requestBody): Promise<any>
}
