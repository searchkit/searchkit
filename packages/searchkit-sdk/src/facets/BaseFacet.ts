import QueryManager, { MixedFilter } from '../core/QueryManager'

export interface FacetOptions {
  size: number
  query: string
}

export interface SelectedFilter {
  type: string
  id: string
  identifier: string
  label: string
  dateMin?: string
  dateMax?: string
  display: string
}

export interface FacetResponseEntry {
  label: string
  count: number
}

export interface FacetResponse {
  identifier: string
  label: string
  type: string
  display: string
  entries?: FacetResponseEntry[]
}

export interface BaseFacet {
  excludeOwnFilters: boolean
  getIdentifier(): string
  getLabel(): string
  getAggregation(overrides: FacetOptions, queryManager: QueryManager): any
  getFilters(filters: Array<MixedFilter>): any
  transformResponse(response: any, queryManager: QueryManager): FacetResponse
  getSelectedFilter(filterSet: any): SelectedFilter
}
