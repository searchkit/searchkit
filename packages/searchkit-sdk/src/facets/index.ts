export { default as RefinementSelectFacet } from './RefinementSelectFacet'
export { default as RangeFacet } from './RangeFacet'
export { default as DateRangeFacet } from './DateRangeFacet'
export { default as HierarchicalMenuFacet } from './HierarchicalMenuFacet'
export { default as MultiQueryOptionsFacet } from './MultiQueryOptionsFacet'
export { default as GeoDistanceOptionsFacet } from './GeoDistanceOptionsFacet'
export * from './BaseFacet'
export * from './VisibilityRules'
