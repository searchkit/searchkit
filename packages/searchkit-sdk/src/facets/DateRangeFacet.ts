import isUndefined from 'lodash/isUndefined'
import { DateRangeFilter } from '../core/QueryManager'
import { BaseFacet } from './BaseFacet'

interface DateRangeFacetConfig {
  identifier: string
  field: string
  label: string
  display?: 'DateRange' | string
}

class DateRangeFacet implements BaseFacet {
  public excludeOwnFilters = true

  constructor(public config: DateRangeFacetConfig) {}
  getLabel(): string {
    return this.config.label
  }

  getIdentifier() {
    return this.config.identifier
  }

  getFilters(filters: Array<DateRangeFilter>) {
    const rangeFilter: { gte?: string; lte?: string } = {}
    if (!isUndefined(filters[0].dateMin)) rangeFilter.gte = filters[0].dateMin
    if (!isUndefined(filters[0].dateMax)) rangeFilter.lte = filters[0].dateMax
    return { range: { [this.config.field]: rangeFilter } }
  }

  getAggregation() {
    return {}
  }

  getSelectedFilter(filterSet) {
    return {
      type: 'DateRangeSelectedFilter',
      id: `${this.getIdentifier()}_${filterSet.dateMin}_${filterSet.dateMax}`,
      identifier: this.getIdentifier(),
      label: this.getLabel(),
      dateMin: filterSet.dateMin,
      dateMax: filterSet.dateMax,
      display: this.config.display || 'DateRangeFacet'
    }
  }

  transformResponse() {
    return {
      identifier: this.getIdentifier(),
      label: this.getLabel(),
      type: 'DateRangeFacet',
      display: this.config.display || 'DateRangeFacet',
      entries: null
    }
  }
}

export default DateRangeFacet
