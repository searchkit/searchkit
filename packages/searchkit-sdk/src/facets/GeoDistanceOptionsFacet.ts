import { ValueFilter } from '../core/QueryManager'
import { BaseFacet } from './BaseFacet'

interface GeoDistanceOptionsFacetConfig {
  identifier: string
  field: string
  origin: string
  unit?: string
  label: string
  display?: string
  ranges: {
    from?: number
    to?: number
    label: string
  }[]
  multipleSelect?: boolean
}

class GeoDistanceOptionsFacet implements BaseFacet {
  public excludeOwnFilters = false
  private unit: string

  constructor(public config: GeoDistanceOptionsFacetConfig) {
    this.excludeOwnFilters = config.multipleSelect
    this.unit = config.unit || 'mi'
  }
  getLabel(): string {
    return this.config.label
  }

  getIdentifier(): string {
    return this.config.identifier
  }

  getFilters(filterValues: ValueFilter[]) {
    const condition = this.excludeOwnFilters ? 'should' : 'must'
    const filters = filterValues.map(({ value }) => {
      const range = this.config.ranges.find((option) => option.label === value)
      return range
    })
    return {
      bool: {
        [condition]: filters.map((range) => ({
          bool: {
            ...(range.from
              ? {
                  must_not: [
                    {
                      geo_distance: {
                        distance: range.from + this.config.unit,
                        [this.config.field]: this.config.origin
                      }
                    }
                  ]
                }
              : {}),
            ...(range.to
              ? {
                  must: [
                    {
                      geo_distance: {
                        distance: range.to + this.config.unit,
                        [this.config.field]: this.config.origin
                      }
                    }
                  ]
                }
              : {})
          }
        }))
      }
    }
  }

  getAggregation() {
    return {
      [this.getIdentifier()]: {
        geo_distance: {
          field: this.config.field,
          origin: this.config.origin,
          unit: this.unit,
          keyed: true,
          ranges: this.config.ranges.map((range) => ({
            from: range.from,
            to: range.to,
            key: range.label
          }))
        }
      }
    }
  }

  getSelectedFilter(filterSet: ValueFilter) {
    return {
      identifier: this.getIdentifier(),
      id: `${this.getIdentifier()}_${filterSet.value}`,
      label: this.getLabel(),
      display: this.config.display || 'ListFacet',
      type: 'GeoDistanceOptionsFacet',
      value: filterSet.value
    }
  }

  transformResponse(response: any) {
    return {
      identifier: this.getIdentifier(),
      label: this.getLabel(),
      display: this.config.display || 'ListFacet',
      type: 'GeoDistanceOptionsFacet',
      entries: this.config.ranges.map((option) => {
        const docCount = response.buckets[option.label]?.doc_count || 0
        return {
          label: option.label,
          count: docCount
        }
      })
    }
  }
}

export default GeoDistanceOptionsFacet
