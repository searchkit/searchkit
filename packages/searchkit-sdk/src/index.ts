import type { SearchInnerHits, SearchRequest } from '@elastic/elasticsearch-types/lib/api/types'
import QueryManager, { MixedFilter, QueryOptions } from './core/QueryManager'
import RequestBodyBuilder from './core/RequestBodyBuilder'
import BaseQuery from './query/BaseQuery'
import { BaseFacet } from './facets/BaseFacet'
import { BaseFilter } from './filters/BaseFilter'
import { VisibleWhenRuleSet } from './facets'
import { SearchkitTransporter, FetchClientTransporter } from './transporters'
import { getAggregationsFromFacets } from './core/FacetsFns'
import {
  SearchkitResponseTransformer,
  ElasticSearchResponseTransformer,
  SearchkitResponse,
  SearchkitHit
} from './transformers'
import { BaseSuggestor } from './suggestors'

export * from './query'
export * from './facets'
export * from './filters'
export * from './transporters'
export * from './suggestors'
export { MixedFilter, QueryOptions, SearchkitResponse, SearchkitHit }

export interface SortingOption {
  id: string
  label: string
  field: any
  defaultOption?: boolean
}

export interface CustomHighlightConfig {
  field: string
  config: any
}

export interface CloudHost {
  id: string
}

export interface BaseConfig {
  host?: string
  cloud?: CloudHost
  index: string
  connectionOptions?: {
    apiKey?: string
    headers?: Record<string, string>
  }
}

export type CollapseInnerHitsConfig = SearchInnerHits

export interface CollapseConfig {
  field: string
  inner_hits?: CollapseInnerHitsConfig | CollapseInnerHitsConfig[]
}

export interface SearchkitConfig extends BaseConfig {
  suggestions?: Array<BaseSuggestor<any>>
  sortOptions?: SortingOption[]
  hits?: {
    fields: string[]
    highlightedFields?: (string | CustomHighlightConfig)[]
  }
  query?: BaseQuery
  facets?: Array<BaseFacet | VisibleWhenRuleSet>
  filters?: Array<BaseFilter>
  collapse?: CollapseConfig
  postProcessRequest?: (body: SearchRequest) => SearchRequest
}

export interface ResultsResolverParameters {
  filters: Array<MixedFilter>
  query: string
}

const getFacets = (
  facets: Array<BaseFacet | VisibleWhenRuleSet> = [],
  queryManager: QueryManager,
  ctx
) =>
  facets.reduce((facetsList, facet) => {
    if (facet instanceof VisibleWhenRuleSet) {
      return [...facetsList, ...facet.getActiveFacets(queryManager, ctx)]
    }
    return [...facetsList, facet]
  }, [])

const createInstance = (
  config: SearchkitConfig,
  transporter?: SearchkitTransporter
): SearchkitRequest => new SearchkitRequest(config, transporter)

export type ResponseRequest = {
  hits?: {
    size?: number
    from?: number
    includeRawHit?: boolean
  }
  facets?:
    | boolean
    | Array<{
        identifier: string
        query?: string
        size?: number
      }>
}

export type BaseFilters = Array<any>

function getSortOption(id, sortOptions: SortingOption[]) {
  const selectedSortOption = sortOptions.find((sortOption) => sortOption.id === id)
  if (!selectedSortOption) {
    throw new Error(`Sort Option: sorting option ${id} not found`)
  }
  return selectedSortOption
}

export class SearchkitRequest {
  private queryManager: QueryManager
  private transporter: SearchkitTransporter
  private transformer: SearchkitResponseTransformer

  constructor(private config: SearchkitConfig, transporter?: SearchkitTransporter) {
    this.queryManager = new QueryManager()
    this.transporter = !transporter ? new FetchClientTransporter(config) : transporter
    this.transformer = new ElasticSearchResponseTransformer()
    this.handleDefaults()
  }

  handleDefaults(): void {
    const defaultSortOptionId =
      this.config.sortOptions && this.config.sortOptions.find((option) => option.defaultOption)
    const defaultSortOption =
      defaultSortOptionId && getSortOption(defaultSortOptionId.id, this.config.sortOptions)
    this.queryManager.setSortBy(defaultSortOption)
  }

  query(query: string): SearchkitRequest {
    this.queryManager.setQuery(query)
    return this
  }

  setFilters(filters: Array<MixedFilter>): SearchkitRequest {
    this.queryManager.setFilters(filters)
    return this
  }

  setQueryOptions(options: QueryOptions): SearchkitRequest {
    this.queryManager.setQueryOptions(options)
    return this
  }

  setSortBy(sortBy: string): SearchkitRequest {
    if (this.config.sortOptions && sortBy) {
      const sortOption = getSortOption(sortBy, this.config.sortOptions)
      this.queryManager.setSortBy(sortOption)
    }
    return this
  }

  async executeSuggestions(query: string): Promise<Array<any>> {
    const suggestions = await Promise.all(
      this.config.suggestions.map((suggestor) => suggestor.getSuggestions(query, this.transporter))
    )
    return suggestions
  }

  async execute(
    responseRequest: ResponseRequest,
    baseFilters: BaseFilters = []
  ): Promise<SearchkitResponse> {
    const partialQueries = []
    const facets = getFacets(this.config.facets, this.queryManager, {})
    let filteredFacets = null
    let overrides = {}

    if (!responseRequest.hits) responseRequest.hits = {}
    if (!responseRequest.hits.size) responseRequest.hits.size = 0
    if (!responseRequest.hits.from) responseRequest.hits.from = 0

    if (responseRequest.facets) {
      if (Array.isArray(responseRequest.facets)) {
        const filteredFacetIdentifiers = responseRequest.facets.map((facet) => facet.identifier)
        filteredFacets =
          facets &&
          facets.filter((facet) => filteredFacetIdentifiers.includes(facet.getIdentifier()))
        overrides = responseRequest.facets.reduce(
          (acc, facet) => ({
            ...acc,
            [facet.identifier]: {
              query: facet.query,
              size: facet.size
            }
          }),
          {}
        )
      }
      const aggs = getAggregationsFromFacets(this.queryManager, overrides, filteredFacets || facets)
      partialQueries.push(aggs)
    }

    const chosenSortOption = this.queryManager.getSortBy()

    partialQueries.push({
      size: responseRequest.hits.size,
      from: responseRequest.hits.from,
      sort: chosenSortOption ? chosenSortOption.field : [{ _score: 'desc' }]
    })

    let skRequestBody = RequestBodyBuilder(
      this.queryManager,
      this.config,
      baseFilters,
      facets,
      partialQueries
    )
    if (this.config.postProcessRequest) {
      skRequestBody = this.config.postProcessRequest(skRequestBody)
    }
    const response = await this.transporter.performRequest(skRequestBody)

    return this.transformer.transformResponse(
      response,
      filteredFacets || facets,
      this.queryManager,
      this.config,
      responseRequest
    )
  }
}

export default createInstance
