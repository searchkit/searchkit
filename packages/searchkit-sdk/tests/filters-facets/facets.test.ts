import nock from 'nock'
import SearchkitRequest from '../../src'
import { RefinementSelectFacet } from '../../src/facets'
import ResultsNoHitsMock from '../__mock-data__/Facets/results-no-hits.json'

describe('Facets', () => {
  it('2 Facets configured, no filters', async () => {
    const request = SearchkitRequest({
      host: 'http://localhost:9200',
      hits: {
        fields: ['facet1']
      },
      facets: [
        new RefinementSelectFacet({
          field: 'writers',
          identifier: 'writers',
          label: 'Writers'
        }),
        new RefinementSelectFacet({
          field: 'actors',
          identifier: 'actors',
          label: 'Actors',
          multipleSelect: false
        })
      ],
      index: 'test'
    })

    const scope = nock('http://localhost:9200')
      .post('/test/_search')
      .reply(200, (uri, body) => {
        expect(body).toMatchSnapshot()
        return ResultsNoHitsMock
      })

    const response = await request.execute({
      facets: true
    })
    expect(response).toMatchSnapshot()
    expect(response.facets).toBeDefined()
    expect(response.facets).toHaveLength(2)
    expect(response.facets.map((f) => f.identifier)).toEqual(['writers', 'actors'])
  })

  it('2 Facets configured, with two filters in each facet', async () => {
    const request = SearchkitRequest({
      host: 'http://localhost:9200',
      hits: {
        fields: []
      },
      facets: [
        new RefinementSelectFacet({
          field: 'writers',
          identifier: 'writers',
          label: 'Writers'
        }),
        new RefinementSelectFacet({
          field: 'actors',
          identifier: 'actors',
          label: 'Actors',
          multipleSelect: false
        })
      ],
      index: 'test'
    })

    request.setFilters([
      { identifier: 'writers', value: 'writer1' },
      { identifier: 'actors', value: 'actors' }
    ])

    const scope = nock('http://localhost:9200')
      .post('/test/_search')
      .reply(200, (uri, body) => {
        expect(body).toMatchSnapshot()
        return ResultsNoHitsMock
      })

    const response = await request.execute({
      facets: true
    })
    expect(response).toMatchSnapshot()
    expect(response.facets).toBeDefined()
    expect(response.facets).toHaveLength(2)
    expect(response.facets.map((f) => f.identifier)).toEqual(['writers', 'actors'])
    expect(response.summary.appliedFilters).toMatchInlineSnapshot(`
      Array [
        Object {
          "display": "ListFacet",
          "id": "writers_writer1",
          "identifier": "writers",
          "label": "Writers",
          "type": "ValueSelectedFilter",
          "value": "writer1",
        },
        Object {
          "display": "ListFacet",
          "id": "actors_actors",
          "identifier": "actors",
          "label": "Actors",
          "type": "ValueSelectedFilter",
          "value": "actors",
        },
      ]
    `)
  })

  it('2 Facets configured, with two filters in each facet, specify a particular facet only to be returned', async () => {
    const request = SearchkitRequest({
      host: 'http://localhost:9200',
      hits: {
        fields: []
      },
      facets: [
        new RefinementSelectFacet({
          field: 'writers',
          identifier: 'writers',
          label: 'Writers'
        }),
        new RefinementSelectFacet({
          field: 'actors',
          identifier: 'actors',
          label: 'Actors',
          multipleSelect: false
        })
      ],
      index: 'test'
    })

    request.setFilters([
      { identifier: 'writers', value: 'writer1' },
      { identifier: 'actors', value: 'actors' }
    ])

    const scope = nock('http://localhost:9200')
      .post('/test/_search')
      .reply(200, (uri, body) => {
        expect(body).toMatchSnapshot()
        return ResultsNoHitsMock
      })

    const response = await request.execute({
      facets: [{ identifier: 'writers' }]
    })
    expect(response).toMatchSnapshot()
    expect(response.facets).toBeDefined()
    expect(response.facets).toHaveLength(1)
    expect(response.facets.map((f) => f.identifier)).toEqual(['writers'])
  })

  it('2 Facets configured, with two filters in each facet, specify both facets to be returned (test multiple works)', async () => {
    const request = SearchkitRequest({
      host: 'http://localhost:9200',
      hits: {
        fields: []
      },
      facets: [
        new RefinementSelectFacet({
          field: 'writers',
          identifier: 'writers',
          label: 'Writers'
        }),
        new RefinementSelectFacet({
          field: 'actors',
          identifier: 'actors',
          label: 'Actors',
          multipleSelect: false
        })
      ],
      index: 'test'
    })

    request.setFilters([
      { identifier: 'writers', value: 'writer1' },
      { identifier: 'actors', value: 'actors' }
    ])

    const scope = nock('http://localhost:9200')
      .post('/test/_search')
      .reply(200, (uri, body) => {
        expect(body).toMatchSnapshot()
        return ResultsNoHitsMock
      })

    const response = await request.execute({
      facets: [{ identifier: 'writers' }, { identifier: 'actors' }]
    })
    expect(response).toMatchSnapshot()
    expect(response.facets).toBeDefined()
    expect(response.facets).toHaveLength(2)
    expect(response.facets.map((f) => f.identifier)).toEqual(['writers', 'actors'])
  })

  it('facet overrides possible via facet array', async () => {
    const request = SearchkitRequest({
      host: 'http://localhost:9200',
      hits: {
        fields: []
      },
      facets: [
        new RefinementSelectFacet({
          field: 'writers',
          identifier: 'writers',
          label: 'Writers'
        }),
        new RefinementSelectFacet({
          field: 'actors',
          identifier: 'actors',
          label: 'Actors',
          multipleSelect: false
        })
      ],
      index: 'test'
    })

    request.setFilters([
      { identifier: 'writers', value: 'writer1' },
      { identifier: 'actors', value: 'actors' }
    ])

    const scope = nock('http://localhost:9200')
      .post('/test/_search')
      .reply(200, (uri, body) => {
        expect(body).toMatchSnapshot()
        return ResultsNoHitsMock
      })

    const response = await request.execute({
      facets: [
        { identifier: 'writers', query: 'writersQuery' },
        { identifier: 'actors', size: 10 }
      ]
    })
    expect(response).toMatchSnapshot()
    expect(response.facets).toBeDefined()
    expect(response.facets).toHaveLength(2)
    expect(response.facets.map((f) => f.identifier)).toEqual(['writers', 'actors'])
  })
})
