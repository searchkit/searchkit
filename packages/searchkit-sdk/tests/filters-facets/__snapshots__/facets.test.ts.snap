// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Facets 2 Facets configured, no filters 1`] = `
Object {
  "_source": Object {
    "includes": Array [
      "facet1",
    ],
  },
  "aggs": Object {
    "facet_bucket_all": Object {
      "aggs": Object {
        "actors": Object {
          "terms": Object {
            "field": "actors",
            "size": 5,
          },
        },
        "writers": Object {
          "terms": Object {
            "field": "writers",
            "size": 5,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "must": Array [],
        },
      },
    },
  },
  "from": 0,
  "size": 0,
  "sort": Array [
    Object {
      "_score": "desc",
    },
  ],
}
`;

exports[`Facets 2 Facets configured, no filters 2`] = `
Object {
  "facets": Array [
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 83,
          "label": "J.J. Abrams",
        },
        Object {
          "count": 74,
          "label": "Jeffrey Lieber",
        },
        Object {
          "count": 73,
          "label": "Damon Lindelof",
        },
        Object {
          "count": 53,
          "label": "James Manos Jr.",
        },
        Object {
          "count": 53,
          "label": "Jeff Lindsay",
        },
      ],
      "identifier": "writers",
      "label": "Writers",
      "type": "RefinementSelectFacet",
    },
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 73,
          "label": "Naveen Andrews",
        },
        Object {
          "count": 56,
          "label": "Jennifer Carpenter",
        },
        Object {
          "count": 56,
          "label": "Michael C. Hall",
        },
        Object {
          "count": 53,
          "label": "Emilie de Ravin",
        },
        Object {
          "count": 42,
          "label": "Jared Padalecki",
        },
      ],
      "identifier": "actors",
      "label": "Actors",
      "type": "RefinementSelectFacet",
    },
  ],
  "hits": Object {
    "items": Array [],
    "page": Object {
      "from": 0,
      "pageNumber": NaN,
      "size": 0,
      "total": 4162,
      "totalPages": Infinity,
    },
  },
  "sortedBy": undefined,
  "summary": Object {
    "appliedFilters": Array [],
    "disabledFilters": Array [],
    "query": "",
    "sortOptions": Array [],
    "total": 4162,
  },
}
`;

exports[`Facets 2 Facets configured, with two filters in each facet 1`] = `
Object {
  "_source": Object {
    "includes": Array [],
  },
  "aggs": Object {
    "facet_bucket_all": Object {
      "aggs": Object {
        "actors": Object {
          "terms": Object {
            "field": "actors",
            "size": 5,
          },
        },
        "writers": Object {
          "terms": Object {
            "field": "writers",
            "size": 5,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "must": Array [
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "writers": "writer1",
                    },
                  },
                ],
              },
            },
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "actors": "actors",
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    },
  },
  "from": 0,
  "post_filter": Object {
    "bool": Object {
      "must": Array [
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "term": Object {
                  "writers": "writer1",
                },
              },
            ],
          },
        },
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "term": Object {
                  "actors": "actors",
                },
              },
            ],
          },
        },
      ],
    },
  },
  "size": 0,
  "sort": Array [
    Object {
      "_score": "desc",
    },
  ],
}
`;

exports[`Facets 2 Facets configured, with two filters in each facet 2`] = `
Object {
  "facets": Array [
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 83,
          "label": "J.J. Abrams",
        },
        Object {
          "count": 74,
          "label": "Jeffrey Lieber",
        },
        Object {
          "count": 73,
          "label": "Damon Lindelof",
        },
        Object {
          "count": 53,
          "label": "James Manos Jr.",
        },
        Object {
          "count": 53,
          "label": "Jeff Lindsay",
        },
      ],
      "identifier": "writers",
      "label": "Writers",
      "type": "RefinementSelectFacet",
    },
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 73,
          "label": "Naveen Andrews",
        },
        Object {
          "count": 56,
          "label": "Jennifer Carpenter",
        },
        Object {
          "count": 56,
          "label": "Michael C. Hall",
        },
        Object {
          "count": 53,
          "label": "Emilie de Ravin",
        },
        Object {
          "count": 42,
          "label": "Jared Padalecki",
        },
      ],
      "identifier": "actors",
      "label": "Actors",
      "type": "RefinementSelectFacet",
    },
  ],
  "hits": Object {
    "items": Array [],
    "page": Object {
      "from": 0,
      "pageNumber": NaN,
      "size": 0,
      "total": 4162,
      "totalPages": Infinity,
    },
  },
  "sortedBy": undefined,
  "summary": Object {
    "appliedFilters": Array [
      Object {
        "display": "ListFacet",
        "id": "writers_writer1",
        "identifier": "writers",
        "label": "Writers",
        "type": "ValueSelectedFilter",
        "value": "writer1",
      },
      Object {
        "display": "ListFacet",
        "id": "actors_actors",
        "identifier": "actors",
        "label": "Actors",
        "type": "ValueSelectedFilter",
        "value": "actors",
      },
    ],
    "disabledFilters": Array [],
    "query": "",
    "sortOptions": Array [],
    "total": 4162,
  },
}
`;

exports[`Facets 2 Facets configured, with two filters in each facet, specify a particular facet only to be returned 1`] = `
Object {
  "_source": Object {
    "includes": Array [],
  },
  "aggs": Object {
    "facet_bucket_all": Object {
      "aggs": Object {
        "writers": Object {
          "terms": Object {
            "field": "writers",
            "size": 5,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "must": Array [
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "writers": "writer1",
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    },
  },
  "from": 0,
  "post_filter": Object {
    "bool": Object {
      "must": Array [
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "term": Object {
                  "writers": "writer1",
                },
              },
            ],
          },
        },
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "term": Object {
                  "actors": "actors",
                },
              },
            ],
          },
        },
      ],
    },
  },
  "size": 0,
  "sort": Array [
    Object {
      "_score": "desc",
    },
  ],
}
`;

exports[`Facets 2 Facets configured, with two filters in each facet, specify a particular facet only to be returned 2`] = `
Object {
  "facets": Array [
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 83,
          "label": "J.J. Abrams",
        },
        Object {
          "count": 74,
          "label": "Jeffrey Lieber",
        },
        Object {
          "count": 73,
          "label": "Damon Lindelof",
        },
        Object {
          "count": 53,
          "label": "James Manos Jr.",
        },
        Object {
          "count": 53,
          "label": "Jeff Lindsay",
        },
      ],
      "identifier": "writers",
      "label": "Writers",
      "type": "RefinementSelectFacet",
    },
  ],
  "hits": Object {
    "items": Array [],
    "page": Object {
      "from": 0,
      "pageNumber": NaN,
      "size": 0,
      "total": 4162,
      "totalPages": Infinity,
    },
  },
  "sortedBy": undefined,
  "summary": Object {
    "appliedFilters": Array [
      Object {
        "display": "ListFacet",
        "id": "writers_writer1",
        "identifier": "writers",
        "label": "Writers",
        "type": "ValueSelectedFilter",
        "value": "writer1",
      },
    ],
    "disabledFilters": Array [
      Object {
        "disabled": true,
        "identifier": "actors",
      },
    ],
    "query": "",
    "sortOptions": Array [],
    "total": 4162,
  },
}
`;

exports[`Facets 2 Facets configured, with two filters in each facet, specify both facets to be returned (test multiple works) 1`] = `
Object {
  "_source": Object {
    "includes": Array [],
  },
  "aggs": Object {
    "facet_bucket_all": Object {
      "aggs": Object {
        "actors": Object {
          "terms": Object {
            "field": "actors",
            "size": 5,
          },
        },
        "writers": Object {
          "terms": Object {
            "field": "writers",
            "size": 5,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "must": Array [
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "writers": "writer1",
                    },
                  },
                ],
              },
            },
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "actors": "actors",
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    },
  },
  "from": 0,
  "post_filter": Object {
    "bool": Object {
      "must": Array [
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "term": Object {
                  "writers": "writer1",
                },
              },
            ],
          },
        },
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "term": Object {
                  "actors": "actors",
                },
              },
            ],
          },
        },
      ],
    },
  },
  "size": 0,
  "sort": Array [
    Object {
      "_score": "desc",
    },
  ],
}
`;

exports[`Facets 2 Facets configured, with two filters in each facet, specify both facets to be returned (test multiple works) 2`] = `
Object {
  "facets": Array [
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 83,
          "label": "J.J. Abrams",
        },
        Object {
          "count": 74,
          "label": "Jeffrey Lieber",
        },
        Object {
          "count": 73,
          "label": "Damon Lindelof",
        },
        Object {
          "count": 53,
          "label": "James Manos Jr.",
        },
        Object {
          "count": 53,
          "label": "Jeff Lindsay",
        },
      ],
      "identifier": "writers",
      "label": "Writers",
      "type": "RefinementSelectFacet",
    },
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 73,
          "label": "Naveen Andrews",
        },
        Object {
          "count": 56,
          "label": "Jennifer Carpenter",
        },
        Object {
          "count": 56,
          "label": "Michael C. Hall",
        },
        Object {
          "count": 53,
          "label": "Emilie de Ravin",
        },
        Object {
          "count": 42,
          "label": "Jared Padalecki",
        },
      ],
      "identifier": "actors",
      "label": "Actors",
      "type": "RefinementSelectFacet",
    },
  ],
  "hits": Object {
    "items": Array [],
    "page": Object {
      "from": 0,
      "pageNumber": NaN,
      "size": 0,
      "total": 4162,
      "totalPages": Infinity,
    },
  },
  "sortedBy": undefined,
  "summary": Object {
    "appliedFilters": Array [
      Object {
        "display": "ListFacet",
        "id": "writers_writer1",
        "identifier": "writers",
        "label": "Writers",
        "type": "ValueSelectedFilter",
        "value": "writer1",
      },
      Object {
        "display": "ListFacet",
        "id": "actors_actors",
        "identifier": "actors",
        "label": "Actors",
        "type": "ValueSelectedFilter",
        "value": "actors",
      },
    ],
    "disabledFilters": Array [],
    "query": "",
    "sortOptions": Array [],
    "total": 4162,
  },
}
`;

exports[`Facets facet overrides possible via facet array 1`] = `
Object {
  "_source": Object {
    "includes": Array [],
  },
  "aggs": Object {
    "facet_bucket_all": Object {
      "aggs": Object {
        "actors": Object {
          "terms": Object {
            "field": "actors",
            "size": 10,
          },
        },
        "writers": Object {
          "terms": Object {
            "field": "writers",
            "include": "([a-zA-Z]+ )+?[wW][rR][iI][tT][eE][rR][sS]Q[uU][eE][rR][yY].*",
            "size": 5,
          },
        },
      },
      "filter": Object {
        "bool": Object {
          "must": Array [
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "writers": "writer1",
                    },
                  },
                ],
              },
            },
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "term": Object {
                      "actors": "actors",
                    },
                  },
                ],
              },
            },
          ],
        },
      },
    },
  },
  "from": 0,
  "post_filter": Object {
    "bool": Object {
      "must": Array [
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "term": Object {
                  "writers": "writer1",
                },
              },
            ],
          },
        },
        Object {
          "bool": Object {
            "must": Array [
              Object {
                "term": Object {
                  "actors": "actors",
                },
              },
            ],
          },
        },
      ],
    },
  },
  "size": 0,
  "sort": Array [
    Object {
      "_score": "desc",
    },
  ],
}
`;

exports[`Facets facet overrides possible via facet array 2`] = `
Object {
  "facets": Array [
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 83,
          "label": "J.J. Abrams",
        },
        Object {
          "count": 74,
          "label": "Jeffrey Lieber",
        },
        Object {
          "count": 73,
          "label": "Damon Lindelof",
        },
        Object {
          "count": 53,
          "label": "James Manos Jr.",
        },
        Object {
          "count": 53,
          "label": "Jeff Lindsay",
        },
      ],
      "identifier": "writers",
      "label": "Writers",
      "type": "RefinementSelectFacet",
    },
    Object {
      "display": "ListFacet",
      "entries": Array [
        Object {
          "count": 73,
          "label": "Naveen Andrews",
        },
        Object {
          "count": 56,
          "label": "Jennifer Carpenter",
        },
        Object {
          "count": 56,
          "label": "Michael C. Hall",
        },
        Object {
          "count": 53,
          "label": "Emilie de Ravin",
        },
        Object {
          "count": 42,
          "label": "Jared Padalecki",
        },
      ],
      "identifier": "actors",
      "label": "Actors",
      "type": "RefinementSelectFacet",
    },
  ],
  "hits": Object {
    "items": Array [],
    "page": Object {
      "from": 0,
      "pageNumber": NaN,
      "size": 0,
      "total": 4162,
      "totalPages": Infinity,
    },
  },
  "sortedBy": undefined,
  "summary": Object {
    "appliedFilters": Array [
      Object {
        "display": "ListFacet",
        "id": "writers_writer1",
        "identifier": "writers",
        "label": "Writers",
        "type": "ValueSelectedFilter",
        "value": "writer1",
      },
      Object {
        "display": "ListFacet",
        "id": "actors_actors",
        "identifier": "actors",
        "label": "Actors",
        "type": "ValueSelectedFilter",
        "value": "actors",
      },
    ],
    "disabledFilters": Array [],
    "query": "",
    "sortOptions": Array [],
    "total": 4162,
  },
}
`;
